File: scripts/R/doubleml_iv_template.R
# scripts/R/doubleml_iv_template.R
# Reusable template for high-dimensional instrumental variables with DoubleML in R.
#
# Use this when IV identification is plausible and the nuisance relationships
# involving covariates X should be estimated flexibly with cross-fitting.
#
# Install manually if needed:
#   install.packages(c("DoubleML", "mlr3", "mlr3learners"))

required <- c("DoubleML", "mlr3", "mlr3learners")
missing <- required[!vapply(required, requireNamespace, logical(1), quietly = TRUE)]
if (length(missing) > 0) {
  stop(
    "Missing packages: ", paste(missing, collapse = ", "),
    "\nInstall with: install.packages(c(",
    paste(sprintf('"%s"', missing), collapse = ", "),
    "))"
  )
}

library(DoubleML)
library(mlr3)
library(mlr3learners)

set.seed(202610)

# ---------------------------------------------------------------------
# USER INPUT SECTION
# ---------------------------------------------------------------------
# df <- read.csv("your_data.csv")
# outcome <- "Y"
# treatment <- "D"
# instrument <- "Z"
# x_cols <- c("X1", "X2", ...)

if (!exists("df")) {
  n <- 1000
  p <- 10
  X <- matrix(rnorm(n * p), nrow = n)
  colnames(X) <- paste0("X", seq_len(p))
  Z <- 0.6 * X[, 1] - 0.3 * X[, 2] + rnorm(n)
  U <- rnorm(n)
  D <- 0.9 * Z + sin(X[, 1]) + 0.4 * X[, 2]^2 + 0.8 * U + rnorm(n)
  Y <- 1.25 * D + cos(X[, 1]) + X[, 2] * X[, 3] + U + rnorm(n)
  df <- data.frame(Y = Y, D = D, Z = Z, X)
}

outcome <- "Y"
treatment <- "D"
instrument <- "Z"
x_cols <- setdiff(names(df), c(outcome, treatment, instrument))

df_iv <- na.omit(df[, c(outcome, treatment, instrument, x_cols)])
cat("Analysis N:", nrow(df_iv), "\n")
cat("Number of covariates:", length(x_cols), "\n\n")

obj_dml_data <- DoubleMLData$new(
  df_iv,
  y_col = outcome,
  d_cols = treatment,
  z_cols = instrument,
  x_cols = x_cols
)

# Linear learners are a transparent starting point.
# Replace with random forest, gradient boosting, glmnet, etc. after checking runtime and tuning.
ml_l <- lrn("regr.lm")  # E[Y | X]
ml_m <- lrn("regr.lm")  # E[Z | X]
ml_r <- lrn("regr.lm")  # E[D | X]

pliv <- DoubleMLPLIV$new(obj_dml_data, ml_l = ml_l, ml_m = ml_m, ml_r = ml_r, n_folds = 5)
pliv$fit()

cat("DoubleML PLIV summary:\n")
print(pliv$summary)
cat("\nConfidence intervals:\n")
print(pliv$confint())

cat("\nInterpretation reminder:\n")
cat("IV-DML helps estimate nuisance functions with cross-fitting; it does not validate the instrument.\n")
cat("You still need relevance, independence, exclusion, and an appropriate IV estimand.\n")

File: scripts/python/doubleml_iv_template.py
#!/usr/bin/env python3
"""scripts/python/doubleml_iv_template.py

Reusable template for high-dimensional instrumental variables with DoubleML.

Use when IV identification is plausible and flexible/high-dimensional adjustment
for X is needed. Cross-fitting helps with nuisance estimation; it does not make
the instrument valid.

Install manually if needed:
    pip install doubleml scikit-learn pandas numpy
"""

from __future__ import annotations

from typing import Iterable
import numpy as np
import pandas as pd


def require_packages(packages: Iterable[str]) -> None:
    missing = []
    for pkg in packages:
        try:
            __import__(pkg)
        except ImportError:
            missing.append(pkg)
    if missing:
        raise SystemExit(
            "Missing packages: "
            + ", ".join(missing)
            + "\nInstall with: pip install "
            + " ".join(missing)
        )


require_packages(["doubleml", "sklearn", "pandas", "numpy"])

import doubleml as dml
from sklearn.base import clone
from sklearn.ensemble import RandomForestRegressor
from sklearn.linear_model import LassoCV


# ----------------------------------------------------------------------
# USER INPUT SECTION
# ----------------------------------------------------------------------
# df = pd.read_csv("your_data.csv")
# outcome = "Y"
# treatment = "D"
# instrument = "Z"
# x_cols = ["X1", "X2", ...]

if "df" not in globals():
    rng = np.random.default_rng(202611)
    n = 1000
    p = 10
    x = rng.normal(size=(n, p))
    cols = [f"X{j+1}" for j in range(p)]
    z = 0.6 * x[:, 0] - 0.3 * x[:, 1] + rng.normal(size=n)
    u = rng.normal(size=n)
    d = 0.9 * z + np.sin(x[:, 0]) + 0.4 * x[:, 1] ** 2 + 0.8 * u + rng.normal(size=n)
    y = 1.25 * d + np.cos(x[:, 0]) + x[:, 1] * x[:, 2] + u + rng.normal(size=n)
    df = pd.DataFrame(x, columns=cols)
    df["Y"] = y
    df["D"] = d
    df["Z"] = z

outcome = "Y"
treatment = "D"
instrument = "Z"
x_cols = [c for c in df.columns if c not in [outcome, treatment, instrument]]

df_iv = df.dropna(subset=[outcome, treatment, instrument] + x_cols).copy()
print(f"Analysis N: {len(df_iv)}, covariates: {len(x_cols)}")

obj_dml_data = dml.DoubleMLData(df_iv, y_col=outcome, d_cols=treatment, z_cols=instrument, x_cols=x_cols)

learner = RandomForestRegressor(
    n_estimators=200,
    max_depth=6,
    min_samples_leaf=5,
    random_state=202611,
    n_jobs=-1,
)

pliv = dml.DoubleMLPLIV(
    obj_dml_data,
    ml_l=clone(learner),  # E[Y | X]
    ml_m=clone(learner),  # E[Z | X]
    ml_r=clone(learner),  # E[D | X]
    n_folds=5,
)
pliv.fit(store_predictions=True)

print("DoubleML PLIV summary:")
print(pliv.summary)
print("\nConfidence intervals:")
print(pliv.confint())

# Optional benchmark with LassoCV.
lasso = LassoCV(cv=5, random_state=202611)
pliv_lasso = dml.DoubleMLPLIV(
    obj_dml_data,
    ml_l=clone(lasso),
    ml_m=clone(lasso),
    ml_r=clone(lasso),
    n_folds=5,
)
pliv_lasso.fit()
print("\nLassoCV benchmark:")
print(pliv_lasso.summary)

print("\nInterpretation reminder:")
print("IV-DML still requires a credible instrument. Report first-stage strength and instrument validity arguments separately.")
